docs/design/provisioning.md

```diff
@@ -0,0 +1,223 @@
+
+> How does a user request Gold/Silver/Bronze?   Using the StorageClass field for "Gold" while requiring the use of VolumeSelector for Silver/Bronze requires a user to know the difference between volumes.
+
+Using just VolumeSelector gives a user a single path to understand.  Admins retain the ability to manually create storage.  Labels on the volume (or the volume creator) are matched with the claim's volume selector.
+
+
+### Other approaches considered:
+
+* Single ConfigMap with each key in Data representing 1 storage class, the value is opaque json.  Good for versioning the blob but it seemed unwieldy.
+* Many ConfigMaps with "storage-class" as a key in the Data.  Difficult to validate unique storage names.
+* Name of ConfigMap as name of storage class. Naming validation is enforced by API server.  Requires PVC.Spec.StorageClass (or similar) with 1:1 mapping of claim to provisioner.
+
+
+## Plugins
+
+Provisioning is implemented as plugins.  A number of standard plugins will be maintained within the Kubernetes codebase, but administrators may require implementations unique to their needs.  Future enhancements can include a DriverProvisioner to allow admins to create volumes that do not exist as plugins in the source tree.  The recent addition of FlexVolume is the prototype for this model.  That driver interface will improve over time and can eventually be made to provision volumes.
+
+The following plugins will be provided by Kubernetes:
+
+1.  EBSProvisioner -- creates a PV specifically for a claim and then creates the EBS volume in AWS.  Can be configured via parameters to create SSD, Provisioned IOPS, and Magnetic disks (EBS's 3 volume types).
+2.  GCEProvisioner -- creates a PV specifically for a claim and then creates the GCEPersistentDisk in GCE.  Can be configured via parameters to create pd-standard and pd-ssd disks.
+3.  CinderProvisioner -- creates a PV specifically for a claim and then creates a Cinder volume in OpenStack.  Can be configured via parameters to create multiple types (TODO: fill in exactly which types are supported)
+4. CephProvisioner -- creates a PV specifically for a claim and then creates the CephVolume in the infrastructure.
+
+### Plugin Parameters
+
+The example above contains `"params": "${OPAQUE-JSON-MAP}"`, which is a string representing a JSON map of parameters.   Each provisionable volume plugin will document and publish the parameters it accepts.
+
+Plugin params can vary by whatever attributes are exposed by a storage provider.  An AWS EBS volume, for example, can be a general purpose SSD (gp2), provisioned IOPs (io1) for speed and throughput, and magnetic (standard) for low cost.  Additionally, EBS volumes can optionally be encrypted.  Configuring each volume type with optional encryption would require 6 distinct provisioners.  Administrators can mix and match any attributes exposed by the provisioning plugin to create distinct storage classes.
+
```

Comment:
Should add in that plugins will be responsible for handling labels as well, and should fail(?) if they are not able to handle all the labels passed in. And that these labels should be propagated to the provisioned PV.